### PR TITLE
change(common/web): keyboard processor package modularization 🧩

### PR DESCRIPTION
**Pull request by jahorton**
Continuing from #7800, this PR helps address #7309 by fully converting our internal `@keymanapp/keyboard-processor` package (and the testing package `@keymanapp/recorder-core`) to ES6 modules - including the former's unit tests!

Setting reviews to "no whitespace-only changes" is _highly_ advised.  As of https://github.com/keymanapp/keyman/pull/7809/commits/7849377e4dcfdf492fce4a9fde9036652e361eb2:  image

(Namespaces were generally the top-level indent; dropping them means removing the indent, hence the huge line-count difference.)

Note that this change removes the `com.keyman.text.Codes` access point for the `Codes` object; conversion to ES6 modules does mean that we'll be dropping the namespaces, hence no more `com`.  After a bit of thought, I figured the ideal spot to move it would be as a property of `KeyboardInterface`, a global object (during operational KeymanWeb) that we must continue to support.  Since even compiled keyboards still reference that... may as well add an access point to it and redirect debug keyboards there for the `Codes` object.
- This may necessitate a corresponding change in Developer; note the altered testing keyboards for a non-minified example of how to handle said scenario.

Took a while to get the unit tests working; by default, ES module compilation doesn't actually care to ensure a smooth Node-based execution experience.  Fortunately, telling `tsc` to use Node's module resolution and relying on `node_modules`-based paths for scripts from linked projects seems to do the trick - both the `esbuild` bundler and `node` execution for unit tests are happy this way.

Finally, the new `index.ts` file may be worth a look - it publishes the 'old' namespace-style organization of `keyboard-processor` and writes it to the global object... so in theory, with a bit more work, we may be able to link higher-level namespaced code with the modularized version seen here.

Now, whether or not _that_ - putting energy into a hybrid state - is worth the effort... is a different question.  I'm not sure how well it'd work with the namespaced projects' setup at the moment.  That said, I'm also not sure how easy it'll be to properly modularize the top-level KMW code... some of that source has interesting cross-effects across files, which may need serious cleanup before modularizing.  I'm honestly more afraid of the latter than the former.

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by jahorton** [comment]
OK... after a bit of work, I've figured out how to use `tsc` to handle declaration emissions.  This works _fantastically_ for development within the repository, but there are issues for supplying declarations to external consumers.  (These are still far less notable than the limitations & issues I ran into with the `dts-bundle-generator` package.)

So, the point of concern, from a `tsc`-based declaration-only build:
```typescript
declare module "text/keyboardProcessor" {
    import type Keyboard from "keyboards/keyboard";
    import KeyEvent from "text/keyEvent";
    import type { MutableSystemStore } from "text/systemStores";
    import type OutputTarget from "text/outputTarget";
    import KeyboardInterface, { VariableStore } from "text/kbdInterface";
    import RuleBehavior from "text/ruleBehavior";
    import { DeviceSpec } from "@keymanapp/web-utils/build/obj/index.js";

    // ...
}
```

Note that final `import` statement:  `tsc` will not bundle in types from other repo-local packages that are considered external to the TS project itself.  As per https://github.com/microsoft/TypeScript/issues/4433#issuecomment-1324138560, this seems to be a limitation caused by relying upon `compilerOptions.paths` in the TS config.

That said, `compilerOptions.paths` is quite convenient for keeping everything properly separated and modularized during development, and it helps facilitate our incremental build setup.  So, I sought out possible ways forward... and landed on something.

**Observations**:

```typescript
declare module "@keymanapp/web-utils/build/obj/index.js" { /* ... */ }
```

Adding the above block will redirect all import references to this module declaration.  That said, unfortunately you can't nest another `declare module` block within that.

But!

```typescript
<reference path="some-declaration-file.d.ts"/>
```

Such statements are totally legal within declaration files and can be used to 'attach' one to another, sort of like a pseudo-import for the referenced file's declarations.

So, the path forward:
1.  Find all "@keymanapp/..." import declarations and duplicate their bundled (lib/) declaration files _next to the main build product's main .d.ts file_.
2. In each local .d.ts file, rename the final `declare module` block to reflect its original location, as with the `"@keymanapp/web-utils/build/obj/index.js"` example above.
3. For each reference to such a .d.ts file, insert `<reference path>` tags at the top of each .d.ts file so that the cross-declaration-file imports may resolve.

Limitations:
1.  If concatenating all .d.ts files into one (as `<reference path>` effectively does), note that each `declare module "___"` name probably needs to be unique across all projects.
    - Fortunately, the primary not-unique example would be the top-level "index" module, which gets renamed in step 2 above.
2. This may require knowledge of the `obj` vs `lib` build product organization scheme when converting obj-based imports to their lib-based declaration after `esbuild`-bundling occurs.

While not _optimal_, it's at least reasonably viable without being too fiddly.

**Comment by mcdurdin** [comment]
> Note that this change removes the `com.keyman.text.Codes` access point for the `Codes` object; conversion to ES6 modules does mean that we'll be dropping the namespaces, hence no more `com`. After a bit of thought, I figured the ideal spot to move it would be as a property of `KeyboardInterface`, a global object (during operational KeymanWeb) that we must continue to support. Since even compiled keyboards still reference that... may as well add an access point to it and redirect debug keyboards there for the `Codes` object.

Ideally, we move `Codes` to common/web/types -- it's a set of constants that is useful across platforms.

> * This may necessitate a corresponding change in Developer; note the altered testing keyboards for a non-minified example of how to handle said scenario.

Not clear what you are saying here -- are you saying we need to change the compiler to emit something different for debug keyboards? Thus, debug keyboards built for Keyman 17+ will have different references to debug keyboards built for earlier versions?

> While not _optimal_, it's at least reasonably viable without being too fiddly. Note that we can also simply... not bother with this for now; the declarations work flawlessly for local development on components (or all of) Keyman Engine for Web. It only matters if and when we publish components / the engine itself to `npm` and/or external developers request type declaration files.

I think we don't bother with this for now. We're not doing npm publishing of KMW at this point, so we can cross this bridge at the time that we do.

**Comment by jahorton** [comment]
> > Note that this change removes the `com.keyman.text.Codes` access point for the `Codes` object; conversion to ES6 modules does mean that we'll be dropping the namespaces, hence no more `com`. After a bit of thought, I figured the ideal spot to move it would be as a property of `KeyboardInterface`, a global object (during operational KeymanWeb) that we must continue to support. Since even compiled keyboards still reference that... may as well add an access point to it and redirect debug keyboards there for the `Codes` object.
> 
> Ideally, we move `Codes` to common/web/types -- it's a set of constants that is useful across platforms.

I've come to that conclusion, too.

> 
> > While not _optimal_, it's at least reasonably viable without being too fiddly. Note that we can also simply... not bother with this for now; the declarations work flawlessly for local development on components (or all of) Keyman Engine for Web. It only matters if and when we publish components / the engine itself to `npm` and/or external developers request type declaration files.
> 
> I think we don't bother with this for now. We're not doing npm publishing of KMW at this point, so we can cross this bridge at the time that we do.

Yep.  Maybe I didn't imply it strongly enough there?

File: common/web/keyboard-processor/src/text/codes.ts [hunk 1]
```diff
@@ -1,103 +1,103 @@
...
-      "CAPS":0x0100,            // CAPITALFLAG
-      "NO_CAPS":0x0200,         // NOTCAPITALFLAG
-      "NUM_LOCK":0x0400,        // NUMLOCKFLAG
-      "NO_NUM_LOCK":0x0800,     // NOTNUMLOCKFLAG
-      "SCROLL_LOCK":0x1000,     // SCROLLFLAG
-      "NO_SCROLL_LOCK":0x2000,  // NOTSCROLLFLAG
-      "VIRTUAL_KEY":0x4000,     // ISVIRTUALKEY
-      "VIRTUAL_CHAR_KEY":0x8000 // VIRTUALCHARKEY // Unused by KMW, but reserved for use by other Keyman engines.
-    },
+let Codes = {
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/text/codes.ts:1]:
```suggestion
const Codes = {
```

File: common/web/keyboard-processor/src/text/codes.ts [hunk 2]
```diff
@@ -1,103 +1,103 @@
...
+      case 'K_CAPS':
+        return true;
+      default:
+        if(Codes.keyCodes[keyID] >= 50000) { // A few are used by `sil_euro_latin`.
+          return true; // is a 'K_' key defined for layer shifting or 'control' use.
+        }
+        // Refer to text/codes.ts - these are Keyman-custom "keycodes" used for
+        // layer shifting keys.  To be safe, we currently let K_TABBACK and
+        // K_TABFWD through, though we might be able to drop them too.
+        let code = Codes[keyID];
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/text/codes.ts:93]:
```suggestion
        const code = Codes[keyID];
```

Might as well tighten this up as we go?

File: common/web/keyboard-processor/src/text/kbdInterface.ts [hunk 3]
```diff
@@ -1,1164 +1,1180 @@
...
+};
 
-  export class KeyboardInterface {
-    static readonly GLOBAL_NAME = 'KeymanWeb';
+export enum SystemStoreIDs {
+  TSS_LAYER = 33,
+  TSS_PLATFORM = 31,
+  TSS_NEWLAYER = 42,
+  TSS_OLDLAYER = 43
+}
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/text/kbdInterface.ts:190]: This ideally belongs in common/web/types longer term

File: common/web/keyboard-processor/tests/cases/basic-engine.js [hunk 4]
```diff
@@ -1,20 +1,18 @@
...
 
-let KeyboardProcessor = require('../../build/index.bundled.js');
-
-// Required initialization setup.
-global.com = KeyboardProcessor.com; // exports all keyboard-processor namespacing.
-
-let KMWRecorder = require('../../../recorder/build/nodeProctor');
+import KeyboardProcessor from '@keymanapp/keyboard-processor/build/obj/text/keyboardProcessor.js';
+import { KeyboardTest } from '@keymanapp/recorder-core/build/obj/index.js';
+import NodeProctor from '@keymanapp/recorder-core/build/obj/nodeProctor.js';
```
Comment (mcdurdin) [inline common/web/keyboard-processor/tests/cases/basic-engine.js:7]: It'd be good to be able to reduce this to:
```suggestion
import KeyboardProcessor from '@keymanapp/keyboard-processor';
import NodeProctor, { KeyboardTest } from '@keymanapp/recorder-core';
```

**Review by mcdurdin — APPROVED** [review]
Only minor things, I think

**Pull request merged**